### PR TITLE
tackle-148-stakeholder-duplication

### PR DESCRIPTION
**Pull request by jonathanvila**
Issue : https://github.com/konveyor/tackle-pathfinder/issues/43

### Features covered
* Avoid Stakeholders and Stakeholdergroups duplicity on updating an assessment

### Unit , Integration and End2End Tests cases covered
* Unit Test ( AssessmentSvcTest.java )
  * given_CreatedAssessment_When_UpdateSeveralTimes_Then_StakeholdersAreNotDuplicated
* End2End Test on minikube with native image ( check_api.sh )
  * 8 Given an assessment, update few values and check return value is 200

### Pre steps Minikube: 
1. Install minikube : https://minikube.sigs.k8s.io/docs/start/
1. Start minikube : minikube start --kubernetes-version=v1.20.2 --cpus 4 --memory 12000
1. Enable ingress addons : minikube addons enable ingress
1. Create tackle namespace
```Shell
kubectl create namespace tackle
``` 
1. Deploy tackle ( to have the keycloak instance running ) on your K8s cluster
```Shell
$ kubectl apply -f  https://raw.githubusercontent.com/konveyor/tackle/main/kubernetes/kubernetes-tackle.yaml -n tackle
```
5. Edit the Ingress object to allow direct request to Pathfinder API
```Shell
$ kubectl edit ingress tackle -n tackle
```
Add this below `paths:`
```Yaml
     - backend:
          service:
            name: tackle-pathfinder
            port:
              number: 8080 
        path: /pathfinder
        pathType: ImplementationSpecific
```
6. Build and push 
```Shell
$ ./mvnw -U -B package -Dquarkus.container-image.push=true -Dquarkus.container-image.group={your quay user} -Dquarkus.container-image.registry=quay.io -Dquarkus.container-image.username={your quay user} -Dquarkus.container-image.password={your quay pwd} -Pnative
```
7. Change the image in the deployment tackle-pathfinder to be the one you have pushed
8. Rollout to update the image
```Shell
$ kubectl rollout restart deployment tackle-pathfinder -n tackle
```

### Test case with Minikube : Do the full API test
```Shell
$ .github/scripts/check_api.sh
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### Pre steps non containerised local test:
1. run keycloak
```Shell
podman run -it --name keycloak --rm \     
            -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin -e KEYCLOAK_IMPORT=/tmp/keycloak/quarkus-realm.json \
            -e DB_VENDOR=h2 -p 8180:8080 -p 8543:8443 -v ./src/main/resources/keycloak:/tmp/keycloak:Z \
            jboss/keycloak:12.0.2
```
1. run postgresql
```Shell
podman run -it \                          
            --name postgres-pathfinder -e POSTGRES_USER=pathfinder \                                               
            -e POSTGRES_PASSWORD=pathfinder -e POSTGRES_DB=pathfinder_db \                              
            -p 5433:5432 postgres:10.6
```
1. run application
```Shell
./mvnw quarkus:dev
```
 
### Test case with local non containerised : Do the full API test
```Shell
$ .github/scripts/check_api.sh localhost:8085 localhost:8180
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### NOTE 
Script Test step number involved in this feature : 8

### CODE REVIEW

**Review by carlosthe19916 — COMMENTED** [review]
- @jonathanvila I've followed all steps described on https://github.com/konveyor/tackle-pathfinder/issues/43 and run it against the current `main` branch and I couldn't reproduce the issue.
- Using my IDE I've removed all the code changes you've made in this PR and then executed the test `given_CreatedAssessment_When_UpdateSeveralTimes_Then_StakeholdersAreNotDuplicated` and the test passed successfully. I've pushed my code to GH and created a PR to your PR so you can see what I'm talking about https://github.com/jonathanvila/tackle-pathfinder/pull/15/files.

My expectation is that If I remove all your code changes then `given_CreatedAssessment_When_UpdateSeveralTimes_Then_StakeholdersAreNotDuplicated` should fail and if I keep all your code changes the same test should pass successfully; otherwise, the test added in this PR is not verifying a real bug.

There is definitely something I'm missing since as I mentioned I couldn't reproduce the bug using `main` code. @jonathanvila Could you please try to remove all your code except your test and verify that the test created is being fired when it should?

**Comment by carlosthe19916** [comment]
@jonathanvila I consider the changes made to replace `==` by `equals` and the new DB INDEX you created is correct; however, I am still not able to reproduce the issue using `main` or your PR. I tried both `native containers` and the quarkus:dev mode against `main` and this PR without success because I can not reproduce the bug.

Since I can not reproduce the bug I can not determine whether the bug has been solved by this PR or not. @mrizzi since you reported the issue associated with this PR could you please double-check the instructions to reproduce the bug are correct or just verify the bug has been fixed by this PR?

**Comment by jonathanvila** [comment]
> @jonathanvila I consider the changes made to replace `==` by `equals` and the new DB INDEX you created is correct; however, I am still not able to reproduce the issue using `main` or your PR. I tried both `native containers` and the quarkus:dev mode against `main` and this PR without success because I can not reproduce the bug.
> 
> Since I can not reproduce the bug I can not determine whether the bug has been solved by this PR or not. @mrizzi since you reported the issue associated with this PR could you please double-check the instructions to reproduce the bug are correct or just verify the bug has been fixed by this PR?

It could be related to this : https://www.baeldung.com/java-compare-long-values
_This is because Java maintains a constant pool for instances of Long between -128 and 127._

I will check the test to force using values higher than 127

File: src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java [hunk 1]
```diff
@@ -169,6 +169,41 @@ public void given_CreatedAssessment_When_UpdateWithStakeholdersNull_Then_ItDoesn
...
+        assessmentDto.setStakeholders(List.of(100L, 200L, 300L));
+        assessmentSvc.updateAssessment(assessment.id, assessmentDto);
+        assessmentUpdated = Assessment.findById(assessment.id);
+        assertThat(assessmentUpdated.stakeholders).extracting(e -> e.stakeholderId).containsExactlyInAnyOrder(100L, 200L, 300L);
+        assertThat(assessmentUpdated.stakeholdergroups).extracting(e -> e.stakeholdergroupId).containsExactlyInAnyOrder(500L, 600L);        
+        
+        assessmentSvc.updateAssessment(assessment.id, assessmentDto);
+        assessmentUpdated = Assessment.findById(assessment.id);
+        assertThat(assessmentUpdated.stakeholders).extracting(e -> e.stakeholderId).containsExactlyInAnyOrder(100L, 200L, 300L);
+        assertThat(assessmentUpdated.stakeholdergroups).extracting(e -> e.stakeholdergroupId).containsExactlyInAnyOrder(500L, 600L);
```
Comment (carlosthe19916) [inline src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:206]: @jonathanvila What about using the special Long numbers in the test itself? for instance:

```suggestion
        assertThat(assessmentUpdated.stakeholdergroups).extracting(e -> e.stakeholdergroupId).containsExactlyInAnyOrder(500L, 600L);
        
        // Using special Long numbers
        assessmentDto.setStakeholders(List.of(98333L, 98222L, 98111L));
        assessmentDto.setStakeholderGroups(List.of(9877L, 9844L));
        assessmentSvc.updateAssessment(assessment.id, assessmentDto);
        assessmentUpdated = Assessment.findById(assessment.id);
        assertThat(assessmentUpdated.stakeholders).extracting(e -> e.stakeholderId).containsExactlyInAnyOrder(98333L, 98222L, 98111L);
        assertThat(assessmentUpdated.stakeholdergroups).extracting(e -> e.stakeholdergroupId).containsExactlyInAnyOrder(9877L, 9844L);
```

Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:206]: well, the problem only seems to happen with numbers below 128. So I could change the 100 to 150 , wdyt ?

Comment (carlosthe19916) [reply src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:206]: As discussed in a meeting internally the value of the numbers are not important at all, what is more important is to be able to see how this test behaves with and without your code changes.

Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:206]: Well, in fact, the value of the numbers is what matters . That's why I mention you suggest to change several numbers, but only the 100 change could affect the result.
I will change this 100 to any value higher than 127.

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java [hunk 2]
```diff
@@ -651,7 +651,7 @@ public void given_ApplicationAssessed_When_CopyAssessmentToAnotherNonAssessedApp
 				.ignoringCollectionOrder()
 				.isEqualTo(assessmentSource);
 	}
-
+	@Test
```
Comment (carlosthe19916) [inline src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:654]: Could you please add an empty line between `}` and `@Test`?

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java [hunk 3]
```diff
@@ -686,7 +686,7 @@ public void given_ApplicationAssessed_When_CopyAssessmentToAnotherAssessedApp_Th
 			.log().all()
 			.statusCode(400);
 	}
-	
+	@Test
```
Comment (carlosthe19916) [inline src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:689]: Same as before comment above

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java [hunk 4]
```diff
@@ -716,12 +716,12 @@ public void given_ApplicationAssessedButDeleted_When_CopyAssessmentToAnotherAsse
...
 			.body(new ApplicationDto(389500L))
 		.when()
-			.post("/assessments/fromAssessmentId=" + header.getId())
+			.post("/assessments?fromAssessmentId=" + header.getId())
 		.then()
 			.log().all()
 			.statusCode(404);
 	}	
-	
+	@Test
```
Comment (carlosthe19916) [inline src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:724]: Same as before comment above

**Review by carlosthe19916 — COMMENTED** [review]
@jonathanvila Finally I was able to reproduce the bug. The bug happens when `stakeholders, stakeholderGroups` values are greater than 127.

Just a very minor change, please see my comments below.

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java (hunk 4 above)
Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:724]: :+1:

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java (hunk 3 above)
Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:689]: :+1:

File: src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java (hunk 2 above)
Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/controllers/AssessmentsResourceTest.java:654]: :+1:

**Comment by carlosthe19916** [comment]
Thanks @jonathanvila . This PR looks good to me.

**Pull request merged**